### PR TITLE
Add Semaphore

### LINKED ISSUE #1088: Semaphore
It is occasionally useful to have semaphore to limit concurrency. This is related to #172 as it serve as an alternative (more low-level) way and is also mentioned in discussion on RW mutex in #94

### PR DESCRIPTION
**Pull request by ndkoval**
An efficient semaphore implementation.

Fixes #1088

### CODE REVIEW

**Comment by fvasco** [comment]
Some cancellation tests can be added:

* A simple test to ensure that cancellation releases the semaphore
* A test to ensure that cancellation does not release semaphore  before that inner block was completed (in case of blocking code/long task)

File: kotlinx-coroutines-core/jvm/test/internal/SegmentQueueTests.kt [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+                    segments[i].removeSegment()
+                }
+            }
+        }.forEach { it.join() }
+        assertEquals(2, q.numberOfSegments)
+    }
+}
+
+@StressCTest
+class SegmentQueueLFTest {
```
Comment (elizarov) [inline kotlinx-coroutines-core/jvm/test/internal/SegmentQueueTests.kt]: Extract `SegmentQueueLFTest` to a separate file (`SegmentQueueLFTest.kt`)

File: kotlinx-coroutines-core/jvm/test/internal/SegmentQueueTests.kt [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+ * the [SegmentBasedQueue] is represented as an infinite array of segments, each stores one element (see [OneElementSegment]).
+ * Both [enqueue] and [dequeue] operations increment the corresponding global index ([enqIdx] for [enqueue] and
+ * [deqIdx] for [dequeue]) and work with the indexed by this counter cell. Since both operations increment the indices
+ * at first, there could be a race: [enqueue] increments [enqIdx], then [dequeue] checks that the queue is not empty
+ * (that's true) and increments [deqIdx], looking into the corresponding cell after that; however, the cell is empty
+ * because the [enqIdx] operation has not been put its element yet. To make the queue non-blocking, [dequeue] can mark
+ * the cell with [BROKEN] token and retry the operation, [enqueue] at the same time should restart as well; this way,
+ * the queue is obstruction-free.
+ */
+private class SegmentBasedQueue<T>(createFirstSegmentLazily: Boolean) : SegmentQueue<OneElementSegment<T>>(createFirstSegmentLazily) {
```
Comment (elizarov) [inline kotlinx-coroutines-core/jvm/test/internal/SegmentQueueTests.kt]: Extract it to a separate file `SegmentBasedQueue.kt` together with its `OneElementSegment` class.

File: kotlinx-coroutines-core/common/src/internal/SegmentQueue.kt [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        // Link `next` and `prev`.
+        next.movePrevToLeft(prev)
+        prev.movePrevNextToRight(next)
+        // Check whether `prev` and `next` are still in the queue
+        // and help with removing them if needed.
+        if (prev.removed)
+            prev.remove()
+        if (next.removed)
+            next.remove()
+    }
```
Comment (elizarov) [inline kotlinx-coroutines-core/common/src/internal/SegmentQueue.kt]: Commenting out these four last lines breaks no test. There are not enough concurrent stress-test to detect any potential bugs in this code.

File: kotlinx-coroutines-core/common/src/sync/Semaphore.kt [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+                                       // and resumes next suspended acquirer if needed.
+        cont as CancellableContinuation<Unit>
+        cont.resume(Unit)
+    }
+}
+
+private class SemaphoreSegment(id: Long, prev: SemaphoreSegment?): Segment<SemaphoreSegment>(id, prev) {
+    private val acquirers = atomicArrayOfNulls<Any?>(SEGMENT_SIZE)
+
+    operator fun get(index: Int): AtomicRef<Any?> = acquirers[index]
```
Comment (elizarov) [inline kotlinx-coroutines-core/common/src/sync/Semaphore.kt]: It is not supported by atomicfu. You cannot have function returning atomicrefs. Make it return value instead and encapsulate CAS with index in a separate method. Rebase to the latest `develop` branch to see those errors. You'll also need to update `atomicfu_version=0.12.6` in `gradle.properties` as it has fixes for arrays support.

File: kotlinx-coroutines-core/common/src/internal/SegmentQueue.kt [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+                startFrom = first!!
+            }
+        }
+        if (startFrom.id > id) return null
+        // Go through `next` references and add new segments if needed,
+        // similarly to the `push` in the Michael-Scott queue algorithm.
+        // The only difference is that `CAS failure` means that the
+        // required segment has already been added, so the algorithm just
+        // uses it. This way, only one segment with each id can be in the queue.
+        var cur: S = startFrom
```
[the remark this answers is not in the document]
Comment (ndkoval) [reply kotlinx-coroutines-core/common/src/internal/SegmentQueue.kt]: Does not work without explicit specification

File: kotlinx-coroutines-core/common/src/sync/Semaphore.kt [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+     * suspending acquirer if there is one at the point of invocation.
+     */
+    public fun release()
+}
+
+/**
+ * Creates new [Semaphore] instance.
+ */
+@Suppress("FunctionName")
+public fun Semaphore(maxPermits: Int, acquiredPermits: Int = 0): Semaphore = SemaphoreImpl(maxPermits, acquiredPermits)
```
[the remark this answers is not in the document]
Comment (ndkoval) [reply kotlinx-coroutines-core/common/src/sync/Semaphore.kt]: Why not `maxPermits`?

**Comment by ndkoval** [comment]
Arrays in `atomicfu` do not work, see https://github.com/Kotlin/kotlinx.atomicfu/pull/65

File: kotlinx-coroutines-core/common/src/sync/Semaphore.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (elizarov) [reply kotlinx-coroutines-core/common/src/sync/Semaphore.kt]: It's both "max" and "initial" (in the most common usage with default of `acquired == 0`), so a simple `permits` is better and it is consistent with general tradition of keeping parameter names short.

File: kotlinx-coroutines-core/common/src/sync/Semaphore.kt [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+        release()
+    }
+}
+
+private class SemaphoreImpl(private val permits: Int, acquiredPermits: Int)
+    : Semaphore, SegmentQueue<SemaphoreSegment>()
+{
+    init {
+        require(permits > 0) { "Semaphore should have at least 1 permit"}
+        require(acquiredPermits in 0..permits) { "The number of acquired permits should be ≥ 0 and ≤ permits" }
```
Comment (elizarov) [inline kotlinx-coroutines-core/common/src/sync/Semaphore.kt]: Please, don't use unicode in message. Write something like "should in 0..permits". Also, the line above is missing space before closing brace.

File: kotlinx-coroutines-core/common/src/sync/Semaphore.kt [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+    override fun newSegment(id: Long, prev: SemaphoreSegment?)= SemaphoreSegment(id, prev)
```
Comment (elizarov) [inline kotlinx-coroutines-core/common/src/sync/Semaphore.kt]: Space between `)` and `=`

**Pull request merged**